### PR TITLE
[RFC] cpickle based dump/load

### PR DESCRIPTION
**Pull request by last-g**
## TL;DR
This is a prototype of a cpickle based dump/load without streaming support. Feedback required.

## cPickle based dump/load

Hello,

I tried to use `joblib.Memory` for some general propose memoization and persistence and I found out that `joblib.dump` and `joblib.load` is quite slow because of using pickle instead of cpickle.

I spent some time investigating the problem and I started with current implementation, ideas from #342 and my own ideas. And the result is this prototype.

### Approach and implementation description

* Idea is not to override `save` and `load_build` methods but provide special `__reduce__` method and constructor function for deserialization.
* It is not possible to access internal pickler/unpickler state with cpickle so we could not reuse the same file stream and have to store all numpy data in dedicated file(s).
* Access to unpickler state is required from array constructor to load numpy objects from a dedicated file because a file path and an open mode (or a file handle) are required to load data and not known at serialization time. That's why threadlocal storage is used as a global variable.
* On python < 3.3 is't not possible to declare `reduce` function outside of a class so we have to use pure python pickler implementation for older python versions.

#### Pros:
 * same features as current dump/load except streaming (single file output)
 * same performance for numpy structures
 * much faster read for non-numpy structures with any python version
 * much faster write for non-numpy structures with python >= 3.3

#### Cons:
 * little bit hacky (I don't see a way to avoid magic with thread local storage for state passing)
 * impossible to save all data into single file (for that we need to hack pickle implementation details and that's impossible for cpickle)

#### Not addressed in prototype:
 * full backward compatibility
 * hashing algorithm also should be optimized for fast `joblib.Memory`

## Benchmarks:

Benchmarks were performed on my laptop with Xorg/KDE running so an error is relatively big. 

### master with python2

```bash

$ python -V
Python 2.7.12+
$ python benchmarks/bench_pickle.py -dl --size 100000
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.653,  0.2759,       nan,       nan,   1.3
       Big dict,       zlib 3,  0.682,  0.3721,       nan,       nan,   0.4
       Big dict,       mmap r,  0.594,  0.2731,       nan,       nan,   1.3
       Big list,          Raw,  0.186,  0.0623,       nan,       nan,   0.2
       Big list,       zlib 3,  0.210,  0.0866,       nan,       nan,   0.0
       Big list,       mmap r,  0.187,  0.0613,       nan,       nan,   0.2
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuestring
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.805,  0.4065,       nan,       nan,   3.1
       Big dict,       zlib 3,  0.962,  0.5572,       nan,       nan,   1.4
       Big dict,       mmap r,  0.802,  0.4078,       nan,       nan,   3.1
       Big list,          Raw,  0.383,  0.1879,       nan,       nan,   2.0
       Big list,       zlib 3,  0.482,  0.2631,       nan,       nan,   0.9
       Big list,       mmap r,  0.385,  0.1876,       nan,       nan,   2.0
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuearray -m
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  6.837,  3.5940,       nan,       nan,  81.7
       Big dict,       zlib 3,  10.106,  5.9666,       nan,       nan,  74.5
       Big list,          Raw,  6.105,  3.3316,       nan,       nan,  80.6
       Big list,       zlib 3,  9.408,  5.1081,       nan,       nan,  73.8
$ python benchmarks/bench_pickle.py -aA
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
      Big array,          Raw,  2.550,  1.6270,       nan,       nan,  762.9
      Big array,       zlib 3,  29.386,  5.9822,       nan,       nan,  723.6
      Big array,       mmap r,  2.248,  0.0016,       nan,       nan,  762.9
   2 big arrays,          Raw,  3.713,  1.7900,       nan,       nan,  839.2
   2 big arrays,       zlib 3,  31.627,  6.7293,       nan,       nan,  795.9
   2 big arrays,       mmap r,  2.512,  0.0018,       nan,       nan,  839.2

```

### prototype with python2

```bash

$ python -V
Python 2.7.12+
$ python benchmarks/bench_pickle.py -dl --size 100000
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.613,  0.0654,       nan,       nan,   1.3
       Big dict,       zlib 3,  0.688,  0.2351,       nan,       nan,   0.4
       Big dict,       mmap r,  0.606,  0.0649,       nan,       nan,   1.3
       Big list,          Raw,  0.194,  0.0151,       nan,       nan,   0.2
       Big list,       zlib 3,  0.214,  0.0600,       nan,       nan,   0.0
       Big list,       mmap r,  0.190,  0.0146,       nan,       nan,   0.2
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuestring
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.820,  0.0975,       nan,       nan,   3.1
       Big dict,       zlib 3,  0.978,  0.3568,       nan,       nan,   1.4
       Big dict,       mmap r,  0.942,  0.1011,       nan,       nan,   3.1
       Big list,          Raw,  0.476,  0.0444,       nan,       nan,   2.0
       Big list,       zlib 3,  0.490,  0.2025,       nan,       nan,   1.0
       Big list,       mmap r,  0.399,  0.0437,       nan,       nan,   2.0
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuearray -m
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  7.605,  1.7739,       nan,       nan,  81.9
       Big dict,       zlib 3,  10.653,  4.0667,       nan,       nan,  73.8
       Big list,          Raw,  6.689,  1.7060,       nan,       nan,  80.8
       Big list,       zlib 3,  9.942,  3.8809,       nan,       nan,  73.4
$ python benchmarks/bench_pickle.py -aA
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
      Big array,          Raw,  2.032,  1.9293,       nan,       nan,  763.0
      Big array,       zlib 3,  32.844,  6.0443,       nan,       nan,  723.6
      Big array,       mmap r,  3.266,  0.0023,       nan,       nan,  763.0
   2 big arrays,          Raw,  2.237,  1.7739,       nan,       nan,  839.2
   2 big arrays,       zlib 3,  34.270,  6.6405,       nan,       nan,  795.9
   2 big arrays,       mmap r,  2.118,  0.0025,       nan,       nan,  839.2

```

### master with python3

```bash

$ python -V
Python 3.5.2+
$ python benchmarks/bench_pickle.py -dl --size 100000
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.752,  0.3601,       nan,       nan,   1.6
       Big dict,       zlib 3,  0.719,  0.4883,       nan,       nan,   0.6
       Big dict,       mmap r,  0.639,  0.3605,       nan,       nan,   1.6
       Big list,          Raw,  0.193,  0.0819,       nan,       nan,   0.2
       Big list,       zlib 3,  0.212,  0.1129,       nan,       nan,   0.0
       Big list,       mmap r,  0.194,  0.0823,       nan,       nan,   0.2
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuestring
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.886,  0.5511,       nan,       nan,   3.7
       Big dict,       zlib 3,  1.023,  0.7266,       nan,       nan,   1.6
       Big dict,       mmap r,  0.880,  0.5467,       nan,       nan,   3.7
       Big list,          Raw,  0.404,  0.2643,       nan,       nan,   2.3
       Big list,       zlib 3,  0.494,  0.3537,       nan,       nan,   1.0
       Big list,       mmap r,  0.404,  0.2642,       nan,       nan,   2.3
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuearray -m
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  6.679,  5.1654,       nan,       nan,  82.0
       Big dict,       zlib 3,  10.805,  7.0672,       nan,       nan,  74.8
       Big list,          Raw,  6.687,  5.2708,       nan,       nan,  80.6
       Big list,       zlib 3,  9.946,  6.2664,       nan,       nan,  73.9
$ python benchmarks/bench_pickle.py -aA
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
      Big array,          Raw,  2.410,  1.8254,       nan,       nan,  762.9
      Big array,       zlib 3,  29.553,  6.1150,       nan,       nan,  723.6
      Big array,       mmap r,  2.714,  0.0015,       nan,       nan,  762.9
   2 big arrays,          Raw,  2.691,  1.8529,       nan,       nan,  839.2
   2 big arrays,       zlib 3,  32.913,  6.7173,       nan,       nan,  795.9
   2 big arrays,       mmap r,  2.396,  0.0018,       nan,       nan,  839.2


```

### ptototype with python3

```bash

$ python -V
Python 3.5.2+
$ python benchmarks/bench_pickle.py -dl --size 100000
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.031,  0.0334,       nan,       nan,   1.6
       Big dict,       zlib 3,  0.062,  0.0573,       nan,       nan,   0.6
       Big dict,       mmap r,  0.036,  0.0406,       nan,       nan,   1.6
       Big list,          Raw,  0.005,  0.0051,       nan,       nan,   0.2
       Big list,       zlib 3,  0.006,  0.0053,       nan,       nan,   0.0
       Big list,       mmap r,  0.004,  0.0047,       nan,       nan,   0.2
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuestring
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  0.061,  0.0565,       nan,       nan,   3.7
       Big dict,       zlib 3,  0.167,  0.0749,       nan,       nan,   1.6
       Big dict,       mmap r,  0.074,  0.0585,       nan,       nan,   3.7
       Big list,          Raw,  0.028,  0.0220,       nan,       nan,   2.3
       Big list,       zlib 3,  0.088,  0.0366,       nan,       nan,   1.0
       Big list,       mmap r,  0.023,  0.0183,       nan,       nan,   2.3
$ python benchmarks/bench_pickle.py -dl --size 100000 --valuearray -m
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
       Big dict,          Raw,  1.297,  2.1245,       nan,       nan,  82.2
       Big dict,       zlib 3,  4.225,  3.0151,       nan,       nan,  74.1
       Big list,          Raw,  1.355,  2.3261,       nan,       nan,  80.8
       Big list,       zlib 3,  4.196,  3.2450,       nan,       nan,  73.4
$ python benchmarks/bench_pickle.py -aA
        Dataset,     strategy,  write,    read, mem_write,  mem_read,  disk
      Big array,          Raw,  1.778,  1.7705,       nan,       nan,  763.0
      Big array,       zlib 3,  30.877,  6.2192,       nan,       nan,  723.6
      Big array,       mmap r,  2.105,  0.0022,       nan,       nan,  763.0
   2 big arrays,          Raw,  2.007,  1.8471,       nan,       nan,  839.2
   2 big arrays,       zlib 3,  32.511,  7.3159,       nan,       nan,  795.9
   2 big arrays,       mmap r,  1.871,  0.0023,       nan,       nan,  839.2


```

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
The loss of single file/streaming is a bit of a big hit for us: in terms
of vision it was important because it reduced the memory overhead and
because it enables saving to filehandles, which is important for cloud
storage.

**Comment by last-g** [comment]
I totally understand this disadvantage that's why it's only a prototype. Actually, I want start discussion how to improve serialization/deserialization speed for atleast several usecases (`Parallel`, `Memory`) or maybe you already have vision how to improve performance and I could help.

Points from my side:
* I don't see memory overhead with two files. I copied all optimizations from original implementation. Am I missing something?
* For `Memory`  it's quite important to be fast with general structures and this implementation could be used as possible backend for storing results. I've tried to use `Memory` for caching some big JSON API and found out that time needed to load data from cache is 5 times bigger than to load data from remote api.
* For `Parallel` usecase is critical to have fast serialization/deserialization of general structures because we are using memory as storage and io speed is not a bottleneck but serialization/deserialization is. Also lack of streaming support is not a problem with `Parallel` usecase.

**Comment by aabadie** [comment]
> I don't see memory overhead with two files. I copied all optimizations from original implementation. Am I missing something?

The memory was mainly occurring when using compression (because the data was compressed in a buffer in memory).

The speed limitations with general data structures (dict, list, etc) were already known (see the discussion in #260). By the way, your prototype is also faster with numpy arrays.
But as said by @GaelVaroquaux, the single file persistence is important when one wants to persist data on remote file storage (S3, HFDS of via sockets) and this is one of the goal of Joblib.

You say it's impossible to pickle in a single file. Can you explain why in details ?
I think your changes are worth having a look but the single file persistence is mandatory. Also, looking briefly at your changes, I had a hard time understand what were your effective changes and what is the actual code in Joblib. Can you provide the prototype in the same file (and don't use duplicates _v2) ? I also noticed that you updated the socket test. This should be part of a different PR if it's relevant.
